src/evict/evict_lru.c

```diff
@@ -1912,8 +1921,11 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
             __wt_cache_read_gen_new(session, page);
 
         /* Pages being forcibly evicted go on the urgent queue. */
-        if (modified &&
-          (page->read_gen == WT_READGEN_OLDEST || page->memory_footprint >= btree->splitmempage)) {
+        if ((modified &&
+              (page->read_gen == WT_READGEN_OLDEST ||
+                page->memory_footprint >= btree->splitmempage)) ||
+          (WT_IS_HS(btree) && __wt_cache_almost_full(session) &&
+            __wt_cache_bytes_plus_overhead(cache, cache->bytes_hs) >= conn->cache_size / 4)) {
```

Comment:
This is quite complex and appears twice. I'd rather it was in a function, something like `__wt_cache_hs_dominant` or similar.